interfaces/sourcegen/sourcegen/_types.py

```diff
@@ -0,0 +1,72 @@
+# This file is part of Cantera. See License.txt in the top-level directory or
+# at https://cantera.org/license.txt for license and copyright information.
+
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass, fields
+import os
+
+
+def _unpack(obj):
+    return (getattr(obj, field.name) for field in fields(obj))
+
+
+@dataclass
+class Param:
+    """ Represents a function parameter """
+
+    p_type: str
+    name: str
+
+
+    @staticmethod
+    def parse(c_param: str):
+        parts = c_param.strip().rsplit(' ', 1)
+        if len(parts) == 2:
+            return Param(*parts)
```

Comment:
My thinking here is the __init__ method for “record”-style classes should be simple setters (as @dataclass generates), and then provide static factories as needed. This is 100% how I would do it in other langs that support "records".

Reply:
I haven't looked at the new commits, but in this case the `__post_init__` method is precisely for this use case in dataclasses, it's why it exists. And in Python, anyways, the factor is provided by the class method, not a static method. That way, subclasses can still take advantage of them and return the subclass type.

```python
>>> class A:
...     @classmethod
...     def create_me(cls, param):
...         return cls(param)
...     def __init__(self, param):
...         self.param = param
...
>>> class B(A):
...     def __init__(self, param):
...         self.b_param = param
...
>>> b = B.create_me("string")
>>> b.param
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'B' object has no attribute 'param'
>>> b.b_param
'string'
```

With a `staticmethod`, you'd always get the superclass type (or really, whichever class was hardcoded as the return value). Hence, the general practice is to use `classmethod` for factories like this.

Reply:
I see how that could be useful, but in this case I'm not envisioning a polymorphic factory that just returns whatever `cls()` it happens to be called from. I think of a parse method as returning a single type or possibly one of several options based on the content being parsed; again, not based type calling it at that moment.

Consider my current setup where I have class `Func` and its child `CsFunc`. `Func` can be parsed out of a C header file, whereas `CsFunc` cannot. But I also understand that child classes just inherit the base classes’s statics, so it might be surprising to a user that `CsFunc.Parse()` returns a `Func`, or is even callable at all (This is of course diffent from C\[++/#] where statics are not inherited [yet, although C# is set to introduce overridable statics in .NET7 / C#11]). I see two options that I think make sense, then:
- `Func.Parse()` `raise`s when `type(cls)` cannot be handled. But this makes these weird inversions of base class knowing about subclass that I think OO-purists argues against (but see also `pathlib.Path()` which always returns a more derived type based on platform and also let’s not get too concerned about following OO dogma)
- Create a separate parsing module with appropriate functions, as in `_parsing.parse_func()`, `_parsing.parse_param()`, etc.

Reply:
Follow up, does using the class method give me this escape hatch? Is it idiomatic?
```
def parse(cls, text):
    if type(cls) != Me:
        raise TypeError
```